sklearn/cluster/optics_.py

```diff
@@ -719,14 +716,12 @@ def _correct_predecessor(reachability_plot, predecessor, ordering, s, e):
        "Improving the Cluster Structure Extracted from OPTICS Plots." Proc. of
        the Conference "Lernen, Wissen, Daten, Analysen" (LWDA) (2018): 318-329.
     """
-
     while s < e:
         if reachability_plot[s] > reachability_plot[e]:
             return s, e
-        p_e = predecessor[e]
-        for i in range(s, e):
-            if p_e == ordering[i]:
-                return s, e
+        p_e = ordering[predecessor[e]]
```

Comment:
This does not look right to me.
And that is why we need a proper _test_.
`predecessor` supposedly contains object indexes in the original data order. `ordering` is a map of cluster-order-index to original data indexes. Using an object index as index to an order-indexed array is wrong (this really should be type safe, not everything int-based!).

Reply:
I'm even happy to have the `precedecessor_correction` out until we actually find a test for it. You have some examples in the paper, do you happen to have them and could share?

Reply:
If `ordering` were the inverse, a map of objects to their position in the cluster order, then it would be fine. So IMHO @qinhanmin2014's `179260591edba5e3b3cf71ec709bf7157309ed76` should be reverted.

Reply:
> ordering is a map of cluster-order-index to original data indexes.

predecessor contains object index in the data order, so we use ordering to map it from the data order to the RD plot order?

Reply:
It may be possible to vectorize this though. `(ordering[s:e] == p_e).any()` or something like this.
But as this is expected to run only for a few points per cluster, the performance impact should be negligible.

Reply:
> > ordering is a map of cluster-order-index to original data indexes.
> 
> predecessor contains object index in the data order, so we use ordering to map it from the data order to the RD plot order?

@qinhanmin2014 `ordering` is _not_ in data order.

Reply:
> @qinhanmin2014 ordering is not in data order.

I see, apologies for the stupid mistake.

Reply:
@adrinjalali please revert https://github.com/scikit-learn/scikit-learn/pull/12077/commits/179260591edba5e3b3cf71ec709bf7157309ed76, thanks

Reply:
@adrinjalali the data set used there is 10k points. That would be overkill for a unit test. There are the ToS of Twitter in the way, too, as these coordinates were extracted from Tweets. I had a simpler ~10-points example once, but I already couldn't find it anymore when writing that paper...
This seems likely to happen with nested clusters; so a suitable unit test could arise from nesting Gaussians.

Reply:
> this really should be type safe, not everything int-based

That would be a very nice benefit of type safety, if used to that extent!